smolyay/surrogate.py

```diff
@@ -1,546 +1,717 @@
-            Method of solving linear equations:
-                'lu': lower upper decomposition
-                'inv': solve for coefficients through x = A^{-1}B.
-                'lstsq': least-square solution.
+
+        Returns
+        -------
+        self : object
+            BaseSurrogate class instance
 
         Raises
         ------
         IndexError
-            Data should be of length of the grid points.
+            X must be 2D array with shape (n_samples, num_dimensions).
+        IndexError
+            y must be 1D array with shape (n_samples,).
         ValueError
-            Solver should be selected between defined methods.
+            X must lie in domain of surrogate.
         """
-        self._data = numpy.array(data, ndmin=1)
-        if self._data.shape != (len(self.grid.points),):
-            raise IndexError("Data must be same length as grid points.")
-
-        # make basis matrix
-        points, basis_functions = numpy.array(self.grid.points), self.grid.basis_functions
-        basis_matrix = numpy.zeros((len(points), len(points)))
-        for j, basis in enumerate(basis_functions):
-            if self.dimension > 1:
-                value = numpy.prod([basis[i](points[...,i]) for i in range(len(basis))],axis=0)
+        # reset constant
+        self._integration_constant = 0
+        if not self._terms_constructed_cache:
+            self._create_terms()
+            self._terms_constructed_cache = True
+
+        # get points
+        if isinstance(
+            X,
+            (
+                smolyay.samples.UnidimensionalPointSet,
+                smolyay.samples.MultidimensionalPointSet,
+            ),
+        ):
+            X = X.points
+
+        # validate data inputs
+        X = numpy.array(X,ndmin=2,copy=None)
+        self._in_domain(X)
+
+        y = numpy.asarray(y)
+        if y.shape != (X.shape[0],) and y.shape != (X.shape[0], 1):
+            raise IndexError("Must be 1D array with shape (n_samples,).")
+
+        # create lookup table and solve for all the basis functions
+        lookup_table = [
+            self.basis_sets[dim](X[:, dim], self.domain[dim])
+            for dim in range(self.num_dimensions)
+        ]
+
+        # create basis matrix
+        if any(
+            any(bf._is_complex for bf in basis_set) for basis_set in self.basis_sets
+        ):
+            basis_matrix = numpy.zeros(
+                (len(X), len(self._index_combinations)), dtype=complex
+            )
+        else:
+            basis_matrix = numpy.zeros((len(X), len(self._index_combinations)))
+
+        # use lookup table to solve for each term
+        for term, ic in enumerate(self._index_combinations):
+            basis_matrix[:, term] = numpy.prod(
+                [lookup_table[dim][ic[dim], :] for dim in range(len(ic))], axis=0
+            )
+
+        # solve for coefficients
+        if self.regularization is None or numpy.any(numpy.iscomplex(basis_matrix)):
```

Comment:
I think you want like this:
```suggestion
        if self.regularization is None or numpy.iscomplexobj(basis_matrix):
```
There is a subtle different between `iscomplexobj` and `iscomplex` if you check the docs